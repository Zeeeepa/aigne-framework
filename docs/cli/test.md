# Testing

Use AIGNE CLI to run agent tests.

## `aigne test` Command

Run tests in the specified agent directory.

```bash
aigne test [options]
```

### Options

* `--path <path>`: Path to the agent directory (defaults to current directory `.`)
* `--help`: Show command help

### Examples

#### Run tests in current directory

```bash
aigne test
```

#### Run tests in specific directory

```bash
aigne test --path ./my-agents
```

### Test File Structure

The AIGNE testing system automatically finds and runs test files in your project. Make sure your test files follow the correct naming conventions and structure.

### Best Practices

1. **Organize test files**: Place test files in appropriate directory structures
2. **Write clear test cases**: Ensure each test case has a clear purpose and expected outcome
3. **Use descriptive test names**: Make test names clearly express what the test covers
4. **Run tests regularly**: Run tests regularly during development to ensure code quality
