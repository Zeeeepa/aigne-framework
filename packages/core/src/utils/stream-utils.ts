import equal from "fast-deep-equal";
import {
  type AgentProcessAsyncGenerator,
  type AgentResponseChunk,
  type AgentResponseStream,
  isAgentResponseDelta,
  isEmptyChunk,
  type Message,
} from "../agents/agent.js";
import { isNonNullable, isRecord, omitBy, type PromiseOrValue } from "./type-utils.js";
import "./stream-polyfill.js";
import type { ReadableStreamDefaultReadResult } from "bun";

export function objectToAgentResponseStream<T extends Message>(json: T): AgentResponseStream<T> {
  if (!isRecord(json)) {
    throw new Error(
      `expect to return a record type such as {result: ...}, but got (${typeof json}): ${json}`,
    );
  }

  return new ReadableStream({
    pull(controller) {
      controller.enqueue({ delta: { json } });
      controller.close();
    },
  });
}

export function mergeAgentResponseChunk<T extends Message>(
  output: T,
  chunk: AgentResponseChunk<T>,
) {
  if (isAgentResponseDelta(chunk)) {
    if (chunk.delta.text) {
      for (const [key, text] of Object.entries(chunk.delta.text)) {
        const original = output[key] as string | undefined;
        const t = (original || "") + (text || "");
        if (t) Object.assign(output, { [key]: t });
      }
    }

    if (chunk.delta.json) {
      Object.assign(
        output,
        omitBy(chunk.delta.json, (v) => v === undefined),
      );
    }
  }

  return output;
}

export async function agentResponseStreamToObject<T extends Message>(
  stream: AgentResponseStream<T> | AgentProcessAsyncGenerator<T>,
): Promise<T> {
  const json: T = {} as T;

  if (stream instanceof ReadableStream) {
    for await (const value of stream) {
      mergeAgentResponseChunk(json, value);
    }
  } else {
    for (;;) {
      const chunk = await stream.next();
      if (chunk.value) {
        if (chunk.done) {
          Object.assign(json, chunk.value);
        } else {
          mergeAgentResponseChunk(json, chunk.value);
        }
      }
      if (chunk.done) break;
    }
  }

  return json;
}

export function asyncGeneratorToReadableStream<T extends Message>(
  generator: AgentProcessAsyncGenerator<T>,
): AgentResponseStream<T> {
  return new ReadableStream({
    async pull(controller) {
      try {
        const chunk = await generator.next();
        if (chunk.value) {
          if (chunk.done) {
            controller.enqueue({ delta: { json: chunk.value } });
          } else {
            controller.enqueue(chunk.value);
          }
        }

        if (chunk.done) {
          controller.close();
        }
      } catch (error) {
        controller.error(error);
      }
    },
  });
}

export function onAgentResponseStreamEnd<T extends Message>(
  stream: AgentResponseStream<T>,
  options?: {
    onChunk?: (
      chunk: AgentResponseChunk<T>,
    ) => PromiseOrValue<AgentResponseChunk<T> | undefined | void>;
    onResult?: (result: T) => PromiseOrValue<Partial<T> | undefined | void>;
    onError?: (error: Error) => PromiseOrValue<Error>;
  },
): AgentResponseStream<T> {
  const json: T = {} as T;
  const reader = stream.getReader();

  return new ReadableStream({
    async pull(controller) {
      try {
        while (true) {
          const { value, done } = await reader.read();

          if (done) {
            const result = (await options?.onResult?.(json)) ?? json;

            if (result && !equal(result, json)) {
              let chunk: AgentResponseChunk<T> = { delta: { json: result } };
              chunk = (await options?.onChunk?.(chunk)) ?? chunk;
              controller.enqueue(chunk);
            }

            controller.close();

            return;
          }

          mergeAgentResponseChunk(json, value);

          const chunk = (await options?.onChunk?.(value)) ?? value;
          if (!isEmptyChunk(chunk)) {
            controller.enqueue(chunk);
            break;
          }
        }
      } catch (error) {
        controller.error((await options?.onError?.(error)) ?? error);
      }
    },
  });
}

export function isAsyncGenerator<T extends AsyncGenerator>(
  value: AsyncGenerator | unknown,
): value is T {
  return typeof value === "object" && value !== null && Symbol.asyncIterator in value;
}

export async function* arrayToAgentProcessAsyncGenerator<T extends Message>(
  chunks: (AgentResponseChunk<T> | Error)[],
  result?: Partial<T>,
): AgentProcessAsyncGenerator<T> {
  for (const chunk of chunks) {
    if (chunk instanceof Error) throw chunk;

    yield chunk;
  }
  if (result !== undefined) return result;
}

export function arrayToReadableStream<T>(chunks: (T | Error)[]): ReadableStream<T> {
  const list = [...chunks];

  return new ReadableStream({
    pull(controller) {
      const item = list.shift();
      if (!item) {
        controller.close();
        return;
      }

      if (item instanceof Error) {
        controller.error(item);
        return;
      }

      controller.enqueue(item);
    },
  });
}

export async function readableStreamToArray<T>(
  stream: ReadableStream<T>,
  options: { catchError: true },
): Promise<(T | Error)[]>;
export async function readableStreamToArray<T>(
  stream: ReadableStream<T>,
  options?: { catchError?: boolean },
): Promise<T[]>;
export async function readableStreamToArray<T>(
  stream: ReadableStream<T>,
  options?: { catchError?: boolean },
): Promise<(T | Error)[]> {
  const result: T[] = [];
  try {
    for await (const value of stream) {
      result.push(value);
    }
  } catch (error) {
    if (!options?.catchError) throw error;

    result.push(error);
  }
  return result;
}

export function stringToAgentResponseStream(
  str: string,
  key: "text" | string = "text",
): AgentResponseStream<Message> {
  const segmenter = new Intl.Segmenter(undefined, { granularity: "word" });
  const segments = segmenter.segment(str);

  return arrayToReadableStream(
    Array.from(segments).map((segment) => ({ delta: { text: { [key]: segment.segment } } })),
  );
}

export function toReadableStream(stream: NodeJS.ReadStream) {
  return new ReadableStream<Uint8Array>({
    start(controller) {
      const onData = (chunk: Buffer) => {
        controller.enqueue(new Uint8Array(chunk));
      };

      const onEnd = () => {
        cleanup();
        controller.close();
      };

      const onError = (err: Error) => {
        cleanup();
        controller.error(err);
      };

      function cleanup() {
        stream.off("data", onData);
        stream.off("end", onEnd);
        stream.off("error", onError);
      }

      stream.on("data", onData);
      stream.on("end", onEnd);
      stream.on("error", onError);
    },
  });
}

export async function readAllString(stream: NodeJS.ReadStream | ReadableStream): Promise<string> {
  return (
    await readableStreamToArray(
      (stream instanceof ReadableStream ? stream : toReadableStream(stream)).pipeThrough(
        new TextDecoderStream(),
      ),
    )
  ).join("");
}

export function mergeReadableStreams<T1, T2>(
  s1: ReadableStream<T1>,
  s2: ReadableStream<T2>,
): ReadableStream<T1 | T2>;
export function mergeReadableStreams(
  ...streams: (ReadableStream<any> | undefined)[]
): ReadableStream<any>;
export function mergeReadableStreams(
  ...streams: (ReadableStream<any> | undefined)[]
): ReadableStream<any> {
  type Reader = {
    reader: ReadableStreamDefaultReader;
    reading?: Promise<{
      result: ReadableStreamDefaultReadResult<any>;
      item: any;
    }>;
  };

  let readers: Reader[] | undefined;

  return new ReadableStream({
    async pull(controller) {
      try {
        readers ??= streams.filter(isNonNullable).map((s) => ({ reader: s.getReader(), data: [] }));

        while (readers.length) {
          const chunk: Awaited<NonNullable<Reader["reading"]>> = await Promise.race(
            readers.map((i) => {
              i.reading ??= i.reader.read().then((result) => ({ result, item: i }));
              return i.reading;
            }),
          );

          if (chunk.result.value) {
            controller.enqueue(chunk.result.value);
            chunk.item.reading = undefined;
            return;
          }

          if (chunk.result.done) {
            readers = readers.filter((i) => i !== chunk.item);
          }
        }

        controller.close();
      } catch (error) {
        controller.error(error);
        if (readers) for (const item of readers) item.reader.releaseLock();
      }
    },
  });
}
