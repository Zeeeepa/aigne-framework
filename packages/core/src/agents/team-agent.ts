import type { Context } from "../aigne/context.js";
import { mergeAgentResponseChunk, readableStreamToAsyncIterator } from "../utils/stream-utils.js";
import { type PromiseOrValue, isEmpty } from "../utils/type-utils.js";
import {
  Agent,
  type AgentOptions,
  type AgentProcessResult,
  type AgentResponseChunk,
  type Message,
} from "./agent.js";

/**
 * Defines the processing modes available for a TeamAgent.
 *
 * The processing mode determines how the agents within a team are executed
 * and how their outputs are combined.
 */
export enum ProcessMode {
  /**
   * Process the agents one by one, passing the output of each agent to the next.
   *
   * In sequential mode, agents execute in order, with each agent receiving the
   * combined output from all previous agents as part of its input.
   */
  sequential = "sequential",

  /**
   * Process all agents in parallel, merging the output of all agents.
   *
   * In parallel mode, all agents execute simultaneously, each receiving the same
   * initial input. Their outputs are then combined based on output key ownership.
   */
  parallel = "parallel",
}

/**
 * Configuration options for creating a TeamAgent.
 *
 * These options extend the base AgentOptions and add team-specific settings.
 */
export interface TeamAgentOptions<I extends Message, O extends Message> extends AgentOptions<I, O> {
  /**
   * The method to process the agents in the team.
   * @default {ProcessMode.sequential}
   */
  mode?: ProcessMode;
}

/**
 * TeamAgent coordinates a group of agents working together to accomplish tasks.
 *
 * A TeamAgent manages a collection of agents (its skills) and orchestrates their
 * execution according to a specified processing mode. It provides mechanisms for
 * agents to work either sequentially (one after another) or in parallel (all at once),
 * with appropriate handling of their outputs.
 *
 * TeamAgent is particularly useful for:
 * - Creating agent workflows where output from one agent feeds into another
 * - Executing multiple agents simultaneously and combining their results
 * - Building complex agent systems with specialized components working together
 *
 * @example
 * Here's an example of creating a sequential TeamAgent:
 * {@includeCode ../../test/agents/team-agent.test.ts#example-team-agent-sequential}
 */
export class TeamAgent<I extends Message, O extends Message> extends Agent<I, O> {
  /**
   * Create a TeamAgent from the provided options.
   *
   * @param options Configuration options for the TeamAgent
   * @returns A new TeamAgent instance
   *
   * @example
   * Here's an example of creating a sequential TeamAgent:
   * {@includeCode ../../test/agents/team-agent.test.ts#example-team-agent-sequential}
   *
   * @example
   * Here's an example of creating a parallel TeamAgent:
   * {@includeCode ../../test/agents/team-agent.test.ts#example-team-agent-parallel}
   */
  static from<I extends Message, O extends Message>(options: TeamAgentOptions<I, O>) {
    return new TeamAgent(options);
  }

  /**
   * Create a new TeamAgent instance.
   *
   * @param options Configuration options for the TeamAgent
   */
  constructor(options: TeamAgentOptions<I, O>) {
    super(options);
    this.mode = options.mode ?? ProcessMode.sequential;
  }

  /**
   * The processing mode that determines how agents in the team are executed.
   *
   * This can be either sequential (one after another) or parallel (all at once).
   */
  mode: ProcessMode;

  /**
   * Process an input message by routing it through the team's agents.
   *
   * Depending on the team's processing mode, this will either:
   * - In sequential mode: Pass input through each agent in sequence, with each agent
   *   receiving the combined output from previous agents
   * - In parallel mode: Process input through all agents simultaneously and combine their outputs
   *
   * @param input The message to process
   * @param context The execution context
   * @returns A stream of message chunks that collectively form the response
   */
  process(input: I, context: Context): PromiseOrValue<AgentProcessResult<O>> {
    switch (this.mode) {
      case ProcessMode.sequential:
        return this._processSequential(input, context);
      case ProcessMode.parallel:
        return this._processParallel(input, context);
    }
  }

  /**
   * Process input sequentially through each agent in the team.
   *
   * This method:
   * 1. Executes each agent in order
   * 2. Passes the combined output from previous agents to the next agent
   * 3. Yields output chunks as they become available
   * 4. Updates the team's agent list with any changes that occurred during processing
   *
   * @param input The message to process
   * @param context The execution context
   * @returns A stream of message chunks from all agents
   *
   * @private
   */
  async *_processSequential(input: I, context: Context): PromiseOrValue<AgentProcessResult<O>> {
    const output: Message = {};

    // Clone the agents to run, so that we can update the agents list during the loop
    const agents = [...this.skills];
    const newAgents: Agent[] = [];

    for (const agent of agents) {
      const [o, transferToAgent] = await context.invoke(
        agent,
        { ...input, ...output },
        { returnActiveAgent: true, streaming: true },
      );

      for await (const chunk of readableStreamToAsyncIterator(o)) {
        yield chunk;
        mergeAgentResponseChunk(output, chunk);
      }
      newAgents.push(await transferToAgent);
    }

    this.skills.splice(0);
    this.skills.push(...newAgents);
  }

  /**
   * Process input in parallel through all agents in the team.
   *
   * This method:
   * 1. Executes all agents simultaneously with the same input
   * 2. Yields combined output chunks
   * 3. Updates the team's agent list with any changes that occurred during processing
   *
   * @param input The message to process
   * @param context The execution context
   * @returns A stream of combined message chunks from all agents
   *
   * @private
   */
  async *_processParallel(input: I, context: Context): PromiseOrValue<AgentProcessResult<O>> {
    const result = await Promise.all(
      this.skills.map((agent) =>
        context.invoke(agent, input, { returnActiveAgent: true, streaming: true }),
      ),
    );

    const streams = result.map((i) => i[0]);

    type Reader = ReadableStreamDefaultReader<AgentResponseChunk<Message>>;
    type ReaderResult = ReadableStreamReadResult<AgentResponseChunk<Message>>;

    type Task = Promise<{ index: number; reader: Reader } & ReaderResult>;

    const read = async (index: number, reader: Reader): Task => {
      const promise = reader.read();
      return promise.then((result) => ({ ...result, reader, index }));
    };

    const tasks = new Map(streams.map((stream, index) => [index, read(index, stream.getReader())]));

    // NOTE: Flag to check if the output key is used by agent at the index,
    const outputKeyUsed = new Map<string, number>();

    while (tasks.size) {
      const { value, done, reader, index } = await Promise.race(tasks.values());
      tasks.delete(index);
      if (!done) {
        tasks.set(index, read(index, reader));
      }

      if (value) {
        let {
          delta: { text, ...delta },
        } = value;

        if (text) {
          for (const key of Object.keys(text)) {
            // the output key is unused, add to map to lock it
            if (!outputKeyUsed.has(key)) {
              outputKeyUsed.set(key, index);
            }
            // the output key is used by the agent at the index, abandon it
            else if (outputKeyUsed.get(key) !== index) {
              delete text[key];
            }
          }
          if (isEmpty(text)) {
            text = undefined;
          }
        }

        if (!isEmpty(delta.json) || !isEmpty(text)) yield { delta: { ...delta, text } };
      }
    }

    const agents = await Promise.all(result.map((i) => i[1]));
    this.skills.splice(0);
    this.skills.push(...agents);
  }
}
