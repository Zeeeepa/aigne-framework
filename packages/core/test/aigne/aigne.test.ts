import { expect, mock, spyOn, test } from "bun:test";
import assert from "node:assert";
import {
  AIAgent,
  AIGNE,
  FunctionAgent,
  type MessageQueueListener,
  UserInputTopic,
  UserOutputTopic,
  createMessage,
} from "@aigne/core";
import { TeamAgent } from "@aigne/core/agents/team-agent.js";
import { stringToAgentResponseStream } from "@aigne/core/utils/stream-utils.js";
import { OpenAIChatModel } from "../_mocks/mock-models.js";

test("AIGNE simple example", async () => {
  // #region example-simple
  const model = new OpenAIChatModel();

  spyOn(model, "process").mockReturnValueOnce(
    Promise.resolve(stringToAgentResponseStream("Hello, How can I assist you today?")),
  );

  // AIGNE: Main execution engine of AIGNE Framework.
  const aigne = new AIGNE({
    model,
  });

  const agent = AIAgent.from({
    name: "chat",
    description: "A chat agent",
  });

  const result = await aigne.invoke(agent, "hello");
  console.log(result); // { $message: "Hello, How can I assist you today?" }

  expect(result).toEqual({ $message: "Hello, How can I assist you today?" });
  // #endregion example-simple
});

test("AIGNE example with streaming response", async () => {
  // #region example-streaming
  const model = new OpenAIChatModel();

  spyOn(model, "process").mockReturnValueOnce(
    Promise.resolve(stringToAgentResponseStream("Hello, How can I assist you today?")),
  );

  // AIGNE: Main execution engine of AIGNE Framework.
  const aigne = new AIGNE({
    model,
  });

  const agent = AIAgent.from({
    name: "chat",
    description: "A chat agent",
  });

  let text = "";

  const stream = await aigne.invoke(agent, "hello", { streaming: true });
  for await (const chunk of stream) {
    if (chunk.delta.text?.$message) text += chunk.delta.text.$message;
  }

  console.log(text); // Output: Hello, How can I assist you today?

  expect(text).toEqual("Hello, How can I assist you today?");

  // #endregion example-streaming
});

test("AIGNE example shutdown", async () => {
  // #region example-shutdown
  const model = new OpenAIChatModel();

  spyOn(model, "process").mockReturnValueOnce(
    Promise.resolve(stringToAgentResponseStream("Hello, How can I assist you today?")),
  );

  // AIGNE: Main execution engine of AIGNE Framework.
  const aigne = new AIGNE({
    model,
  });

  const agent = AIAgent.from({
    name: "chat",
    description: "A chat agent",
  });

  await aigne.invoke(agent, "hello");

  await aigne.shutdown();
  // #endregion example-shutdown
});

test("AIGNE example shutdown by `using` statement", async () => {
  // #region example-shutdown-using
  const model = new OpenAIChatModel();

  spyOn(model, "process").mockReturnValueOnce(
    Promise.resolve(stringToAgentResponseStream("Hello, How can I assist you today?")),
  );

  // AIGNE: Main execution engine of AIGNE Framework.
  await using aigne = new AIGNE({
    model,
  });

  const agent = AIAgent.from({
    name: "chat",
    description: "A chat agent",
  });

  await aigne.invoke(agent, "hello");

  // aigne will be automatically shutdown when exiting the using block

  // #endregion example-shutdown-using
});

test("AIGNE example invoke get an user agent ", async () => {
  // #region example-user-agent
  const model = new OpenAIChatModel();

  spyOn(model, "process")
    .mockReturnValueOnce(
      Promise.resolve(stringToAgentResponseStream("Hello, How can I assist you today?")),
    )
    .mockReturnValueOnce(Promise.resolve(stringToAgentResponseStream("Nice to meet you, Bob!")));

  // AIGNE: Main execution engine of AIGNE Framework.
  const aigne = new AIGNE({
    model,
  });

  const agent = AIAgent.from({
    name: "chat",
    description: "A chat agent",
  });

  const userAgent = aigne.invoke(agent);

  const result1 = await userAgent.invoke("hello");
  console.log(result1); // { $message: "Hello, How can I assist you today?" }

  expect(result1).toEqual({ $message: "Hello, How can I assist you today?" });

  const result2 = await userAgent.invoke("I'm Bob!");
  console.log(result2); // { $message: "Nice to meet you, Bob!" }

  expect(result2).toEqual({ $message: "Nice to meet you, Bob!" });
  // #endregion example-user-agent
});

test("AIGNE example publish message", async () => {
  // #region example-publish-message
  const model = new OpenAIChatModel();

  spyOn(model, "process").mockReturnValueOnce(
    Promise.resolve(stringToAgentResponseStream("Hello, How can I assist you today?")),
  );

  const agent = AIAgent.from({
    name: "chat",
    description: "A chat agent",
    subscribeTopic: "test_topic",
    publishTopic: "result_topic",
  });

  // AIGNE: Main execution engine of AIGNE Framework.
  const aigne = new AIGNE({
    model,
    // Add agent to AIGNE
    agents: [agent],
  });

  const subscription = aigne.subscribe("result_topic");

  aigne.publish("test_topic", "hello");

  const { message } = await subscription;

  console.log(message); // { $message: "Hello, How can I assist you today?" }

  expect(message).toEqual({ $message: "Hello, How can I assist you today?" });
  // #endregion example-publish-message
});

test("AIGNE example subscribe topic", async () => {
  // #region example-subscribe-topic
  const model = new OpenAIChatModel();

  spyOn(model, "process").mockReturnValueOnce(
    Promise.resolve(stringToAgentResponseStream("Hello, How can I assist you today?")),
  );

  const agent = AIAgent.from({
    name: "chat",
    description: "A chat agent",
    subscribeTopic: "test_topic",
    publishTopic: "result_topic",
  });

  // AIGNE: Main execution engine of AIGNE Framework.
  const aigne = new AIGNE({
    model,
    // Add agent to AIGNE
    agents: [agent],
  });

  const unsubscribe = aigne.subscribe("result_topic", ({ message }) => {
    console.log(message); // { $message: "Hello, How can I assist you today?" }

    unsubscribe();
  });

  aigne.publish("test_topic", "hello");

  // #endregion example-subscribe-topic
});

test("AIGNE.invoke with reflection", async () => {
  const plusOne = FunctionAgent.from({
    subscribeTopic: [UserInputTopic, "revise"],
    publishTopic: "review_request",
    process: (input: { num: number }) => ({ num: input.num + 1 }),
  });

  const reviewer = FunctionAgent.from({
    subscribeTopic: "review_request",
    publishTopic: (output) => (output.num > 10 ? UserOutputTopic : "revise"),
    process: ({ num }: { num: number }) => {
      return {
        num,
        approval: num > 10 ? "approve" : "revise",
      };
    },
  });

  const aigne = new AIGNE({ agents: [plusOne, reviewer] });
  aigne.publish(UserInputTopic, { num: 1 });
  const { message: result } = await aigne.subscribe(UserOutputTopic);

  expect(result).toEqual({ num: 11, approval: "approve" });
});

test("AIGNE.shutdown should shutdown all tools and agents", async () => {
  const plus = FunctionAgent.from(({ a, b }: { a: number; b: number }) => ({
    sum: a + b,
  }));

  const agent = AIAgent.from({});

  const aigne = new AIGNE({
    skills: [plus],
    agents: [agent],
  });

  const plusShutdown = spyOn(plus, "shutdown");
  const agentShutdown = spyOn(agent, "shutdown");

  await aigne.shutdown();

  expect(plusShutdown).toHaveBeenCalled();
  expect(agentShutdown).toHaveBeenCalled();
});

test("AIGNE should throw error if reached max agent calls", async () => {
  const plus = FunctionAgent.from(
    async (
      { num, times }: { num: number; times: number },
      { context },
    ): Promise<{ num: number }> => {
      if (times <= 1) {
        return { num: num + 1 };
      }

      return context.invoke(plus, { num: num + 1, times: times - 1 });
    },
  );

  const aigne = new AIGNE({
    limits: {
      maxAgentInvokes: 2,
    },
  });

  expect(aigne.invoke(plus, { num: 0, times: 2 })).resolves.toEqual({ num: 2 });
  expect(aigne.invoke(plus, { num: 0, times: 3 })).rejects.toThrowError(
    "Exceeded max agent invokes 2/2",
  );
});

test("AIGNE should throw error if reached max tokens", async () => {
  const model = new OpenAIChatModel();

  spyOn(model, "process").mockReturnValue({
    text: "hello",
    usage: {
      inputTokens: 100,
      outputTokens: 200,
    },
  });

  const agent = AIAgent.from({});

  const aigne = new AIGNE({
    model,
    limits: {
      maxTokens: 200,
    },
  });

  expect(aigne.invoke(agent, "test")).resolves.toEqual(createMessage("hello"));
  expect(aigne.invoke(TeamAgent.from({ skills: [agent, agent] }), "test")).rejects.toThrow(
    "Exceeded max tokens 300/200",
  );
});

test("AIGNE should throw timeout error", async () => {
  const agent = FunctionAgent.from(async ({ timeout }: { timeout: number }) => {
    await new Promise((resolve) => {
      setTimeout(resolve, timeout);
    });

    return { timeout };
  });

  const aigne = new AIGNE({
    limits: {
      timeout: 200,
    },
  });

  expect(aigne.invoke(agent, { timeout: 100 })).resolves.toEqual({ timeout: 100 });
  expect(aigne.invoke(agent, { timeout: 300 })).rejects.toThrow("AIGNEContext is timeout");
});

test("AIGNEContext should subscribe/unsubscribe correctly", async () => {
  const aigne = new AIGNE({});

  const listener: MessageQueueListener = mock();

  aigne.subscribe("test_topic", listener);

  aigne.publish("test_topic", "hello");
  expect(listener).toBeCalledTimes(1);
  expect(listener).toHaveBeenCalledWith(
    expect.objectContaining({ message: createMessage("hello") }),
  );

  aigne.unsubscribe("test_topic", listener);
  aigne.publish("test_topic", "hello");
  expect(listener).toBeCalledTimes(1);
});

test("AIGNE.invoke support custom user context", async () => {
  const aigne = new AIGNE<{ id: string; name: string }>({});

  const agent = FunctionAgent.from((_, options) => {
    options.context.userContext.newContextProperty = "foo";

    return options.context.invoke(childAgent, {});
  });

  const childAgent = FunctionAgent.from(() => ({
    text: "I am a child agent",
  }));

  const agentProcess = spyOn(agent, "process");
  const childAgentProcess = spyOn(childAgent, "process");

  const result = await aigne.invoke(agent, "hello", {
    userContext: { id: "test_user_id", name: "test_user_name" },
  });
  expect(result).toEqual({ text: "I am a child agent" });
  expect(agentProcess).toHaveBeenLastCalledWith(
    expect.anything(),
    expect.objectContaining({
      context: expect.objectContaining({
        userContext: expect.objectContaining({ id: "test_user_id", name: "test_user_name" }),
      }),
    }),
  );
  expect(childAgentProcess).toHaveBeenLastCalledWith(
    expect.anything(),
    expect.objectContaining({
      context: expect.objectContaining({
        userContext: expect.objectContaining({
          id: "test_user_id",
          name: "test_user_name",
          newContextProperty: "foo",
        }),
      }),
    }),
  );
});

test("AIGNE.publish support custom user context", async () => {
  const agent = FunctionAgent.from({
    publishTopic: "test_topic",
    subscribeTopic: "test_topic1",
    process: () => ({
      text: "I am a test agent",
    }),
  });

  const aigne = new AIGNE({
    model: new OpenAIChatModel(),
    agents: [agent],
  });
  assert(aigne.model);

  const agentProcess = spyOn(agent, "process");
  spyOn(aigne.model, "process").mockReturnValueOnce({
    text: "I am a test agent",
  });
  const result = aigne.subscribe("test_topic");
  aigne.publish("test_topic1", "hello", {
    userContext: { id: "test_user_id", name: "test_user_name" },
  });
  const { message: response } = await result;
  expect(response).toEqual({ text: "I am a test agent" });
  expect(agentProcess).toHaveBeenLastCalledWith(
    expect.anything(),
    expect.objectContaining({
      context: expect.objectContaining({
        userContext: { id: "test_user_id", name: "test_user_name" },
      }),
    }),
  );
});
