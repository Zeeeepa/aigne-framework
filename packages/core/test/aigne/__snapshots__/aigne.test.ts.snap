// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`AIGNE.invoke should respond progressing chunks correctly 1`] = `
[
  {
    "progress": {
      "agent": {
        "name": "chat",
      },
      "event": "agentStarted",
      "input": {
        "message": "hello",
      },
    },
  },
  {
    "progress": {
      "agent": {
        "name": "OpenAIChatModel",
      },
      "event": "agentStarted",
      "input": {
        "messages": [
          {
            "content": "hello",
            "role": "user",
          },
        ],
        "modelOptions": undefined,
        "responseFormat": undefined,
        "toolChoice": undefined,
        "tools": undefined,
      },
    },
  },
  {
    "delta": {
      "text": {
        "message": "Hello",
      },
    },
  },
  {
    "delta": {
      "text": {
        "message": ",",
      },
    },
  },
  {
    "delta": {
      "text": {
        "message": " ",
      },
    },
  },
  {
    "delta": {
      "text": {
        "message": "How",
      },
    },
  },
  {
    "delta": {
      "text": {
        "message": " ",
      },
    },
  },
  {
    "delta": {
      "text": {
        "message": "can",
      },
    },
  },
  {
    "delta": {
      "text": {
        "message": " ",
      },
    },
  },
  {
    "delta": {
      "text": {
        "message": "I",
      },
    },
  },
  {
    "delta": {
      "text": {
        "message": " ",
      },
    },
  },
  {
    "delta": {
      "text": {
        "message": "assist",
      },
    },
  },
  {
    "delta": {
      "text": {
        "message": " ",
      },
    },
  },
  {
    "delta": {
      "text": {
        "message": "you",
      },
    },
  },
  {
    "delta": {
      "text": {
        "message": " ",
      },
    },
  },
  {
    "progress": {
      "agent": {
        "name": "OpenAIChatModel",
      },
      "event": "agentSucceed",
      "output": {
        "text": "Hello, How can I assist you today?",
      },
    },
  },
  {
    "delta": {
      "text": {
        "message": "today",
      },
    },
  },
  {
    "delta": {
      "text": {
        "message": "?",
      },
    },
  },
  {
    "progress": {
      "agent": {
        "name": "chat",
      },
      "event": "agentSucceed",
      "output": {
        "message": "Hello, How can I assist you today?",
      },
    },
  },
]
`;

exports[`AIGNE.invoke should respond progressing chunks (with failed chunks) correctly 1`] = `
[
  {
    "progress": {
      "agent": {
        "name": "chat",
      },
      "event": "agentStarted",
      "input": {
        "message": "hello",
      },
    },
  },
  {
    "progress": {
      "agent": {
        "name": "OpenAIChatModel",
      },
      "event": "agentStarted",
      "input": {
        "messages": [
          {
            "content": "hello",
            "role": "user",
          },
        ],
        "modelOptions": undefined,
        "responseFormat": undefined,
        "toolChoice": undefined,
        "tools": undefined,
      },
    },
  },
  {
    "progress": {
      "agent": {
        "name": "OpenAIChatModel",
      },
      "error": [Error: Test error],
      "event": "agentFailed",
    },
  },
  {
    "delta": {
      "text": {
        "message": "Hello",
      },
    },
  },
  {
    "progress": {
      "agent": {
        "name": "chat",
      },
      "error": [Error: Test error],
      "event": "agentFailed",
    },
  },
  {
    "delta": {
      "text": {
        "message": " world",
      },
    },
  },
  [Error: Test error],
]
`;

exports[`AIGNE.invoke should respond $meta field with usage metrics with {
  streaming: true,
  returnMetadata: true,
} 1`] = `
[
  {
    "delta": {
      "text": {
        "message": "Hello",
      },
    },
  },
  {
    "delta": {
      "text": {
        "message": " world",
      },
    },
  },
  {
    "delta": {
      "json": {
        "$meta": {
          "usage": {
            "agentCalls": 4,
            "inputTokens": 13,
            "outputTokens": 24,
          },
        },
        "message": "Hello world",
      },
    },
  },
]
`;
